File: ios/Classes/PingEventHandler.swift
//
//  PingEventHandler.swift
//  pingidsdk
//
//  Created by Sandro Caseiro on 6/18/20.
//

import Foundation
import PingID_SDK

public class PingEventHandler : NSObject, PingIDDelegate {
    let channel: FlutterMethodChannel
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    public func handleCallback(_ state: PIDFlowState, _ x: Any?, _ error: Error?) {
        switch state {
        case PIDFlowState.trustLevels:
            onPairingOptionsRequired(availableTrustLevels: x as! [UInt])
        default:
            break
        }
    }
    
    public func handleRemoteNotification(_ notificationType: PIDRemoteNotificationType, _ availableTrustLevels: [Any]?,
                                         _ sessionInfo: [AnyHashable : Any]?, _ error: Error?) {
        let info = sessionInfo as? [String: AnyObject]
        switch notificationType {
        case PIDRemoteNotificationType.auth:
            onAuthenticationRequired(info)
        case PIDRemoteNotificationType.done:
            onAuthenticationCompleted(error as NSError?)
        default:
            break
        }
    }
    
    public func onPairingOptionsRequired(availableTrustLevels: [UInt], deviceName: String? = nil) {
        let args: [String: Any?] = [
            PairingOptionsRequiredArgs.AVAILABLE_TRUST_LEVELS.rawValue: availableTrustLevels.map({ (t: UInt) -> String in
                return PIDTrustLevel(rawValue: t)!.stringValue
            }),
            PairingOptionsRequiredArgs.DEVICE_NAME.rawValue: deviceName
        ]
        
        channel.invokeMethod(FlutterMethods.PAIRING_OPTIONS_REQUIRED.rawValue, arguments: args)
    }
    
    public func onPairingCompleted(_ error: NSError?) {
        var errorMessage: String? = nil
        if (error?.domain == "PIDErrorDomain") {
            errorMessage = PIDError(rawValue: UInt(bitPattern: error!.code))?.description
        }

        let args: [String: Any?] = [
            PairingCompletedArgs.ACTION_STATUS.rawValue: error == nil ? PIDActionStatusValues.SUCCESS.rawValue : PIDActionStatusValues.FAILURE.rawValue,
            PairingCompletedArgs.ERROR.rawValue: errorMessage
        ]
        channel.invokeMethod(FlutterMethods.PAIRING_COMPLETED.rawValue, arguments: args)
    }
    
    public func didUntrustDevice() {
        channel.invokeMethod(FlutterMethods.DID_UNTRUST_DEVICE.rawValue, arguments: nil)
    }
    
    public func onAuthenticationRequired(_ data: [String: AnyObject]?) {
        let args: [String: Any?] = [
            AuthenticationRequiredArgs.DATA.rawValue: data
        ]
        channel.invokeMethod(FlutterMethods.AUTHENTICATION_REQUIRED.rawValue, arguments: args)
    }
    
    public func onAuthenticationCompleted(_ error: NSError?) {
        var errorMessage: String? = nil
        if (error?.domain == "PIDErrorDomain") {
            errorMessage = PIDError(rawValue: UInt(bitPattern: error!.code))?.description
        }
        
        let args: [String: Any?] = [
            AuthenticationCompletedArgs.ACTION_STATUS.rawValue: error == nil ? PIDActionStatusValues.SUCCESS.rawValue : PIDActionStatusValues.FAILURE.rawValue,
            AuthenticationCompletedArgs.ACTION.rawValue: error == nil ? PIDActionTypeValues.PIDActionTypeApprove.rawValue : PIDActionTypeValues.PIDActionTypeNone.rawValue,
            AuthenticationCompletedArgs.ERROR.rawValue: errorMessage
        ]
        channel.invokeMethod(FlutterMethods.AUTHENTICATION_COMPLETED.rawValue, arguments: args)
    }
    
    public func onAuthenticationCancelled() {
        channel.invokeMethod(FlutterMethods.AUTHENTICATION_CANCELLED.rawValue, arguments: nil)
    }
    
    public func didRefreshOneTimePasscode(_ oneTimePasscode: String) {
        let args: [String: Any?] = [
            OneTimePasswordChangedArgs.OTP.rawValue: oneTimePasscode
        ]
        channel.invokeMethod(FlutterMethods.ONE_TIME_PASSWORD_CHANGED.rawValue, arguments: args)
    }

    public func authenticationTokenStatus(_ sessionInfo: [AnyHashable : Any], error: Error?) {
        let args: [String: Any?] = [
            AuthenticationTokenStatusArgs.DATA.rawValue: sessionInfo,
            AuthenticationTokenStatusArgs.ERROR.rawValue: (error as NSError?)?.localizedDescription
        ]
        channel.invokeMethod(FlutterMethods.AUTHENTICATION_TOKEN_STATUS.rawValue, arguments: args)
    }
}

File: android/src/main/kotlin/com/sandrocaseiro/pingidsdk/PingEventHandler.kt
package com.sandrocaseiro.pingidsdk

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.sandrocaseiro.pingidsdk.extensions.toMap
import io.flutter.plugin.common.MethodChannel
import pingidsdkclient.DeviceDetails
import pingidsdkclient.PingID

class PingEventHandler(private val channel: MethodChannel) : PingID.PingIdSdkEvents {
    private val mainHandler = Handler(Looper.getMainLooper())

    //Pairing Events
    override fun onPairingOptionsRequired(availableTrustLevels: MutableList<String>?, deviceDetails: DeviceDetails?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.PAIRING_OPTIONS_REQUIRED, mapOf(
                PairingOptionsRequiredArgs.AVAILABLE_TRUST_LEVELS to availableTrustLevels,
                PairingOptionsRequiredArgs.DEVICE_NAME to deviceDetails?.name
            ))
        }
    }

    override fun onPairingOptionsRequiredWithPasscode(
        availableTrustLevels: MutableList<String>?,
        deviceName: String?
    ) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.PAIRING_OPTIONS_REQUIRED_WITH_PASSWORD, mapOf(
                PairingOptionsRequiredWithPasswordArgs.AVAILABLE_TRUST_LEVELS to availableTrustLevels,
                PairingOptionsRequiredWithPasswordArgs.DEVICE_NAME to deviceName
            ))
        }
    }

    override fun onPairingProgress(msg: String?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.PAIRING_PROGRESS, mapOf(
                PairingProgressArgs.MESSAGE to msg
            ))
        }
    }

    override fun onPairingCompleted(
        status: PingID.PIDActionStatus?,
        error: PingID.PIDErrorDomain?
    ) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.PAIRING_COMPLETED, mapOf(
                PairingCompletedArgs.ACTION_STATUS to status.toString(),
                PairingCompletedArgs.ERROR to error.toString()
            ))
        }
    }

    override fun onIgnoreDeviceCompleted(
        status: PingID.PIDActionStatus?,
        error: PingID.PIDErrorDomain?
    ) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.IGNORE_DEVICE_COMPLETED, mapOf(
                IgnoreDeviceCompletedArgs.ACTION_STATUS to status.toString(),
                IgnoreDeviceCompletedArgs.ERROR to error.toString()
            ))
        }
    }

    override fun didUntrustDevice() {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.DID_UNTRUST_DEVICE, null)
        }
    }

    //Authentication Events
    override fun onAuthenticationRequired(data: Bundle?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.AUTHENTICATION_REQUIRED, mapOf(
                AuthenticationRequiredArgs.DATA to data?.toMap()
            ))
        }

    }

    override fun onAuthenticationCompleted(
        status: PingID.PIDActionStatus?,
        action: PingID.PIDActionType?,
        error: PingID.PIDErrorDomain?
    ) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.AUTHENTICATION_COMPLETED, mapOf(
                AuthenticationCompletedArgs.ACTION_STATUS to status.toString(),
                AuthenticationCompletedArgs.ACTION to action.toString(),
                AuthenticationCompletedArgs.ERROR to error.toString()
            ))
        }
    }

    override fun onAuthenticationCancelled() {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.AUTHENTICATION_CANCELLED, null)
        }
    }

    override fun onOneTimePasscodeChanged(newOneTimePasscode: String?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.ONE_TIME_PASSWORD_CHANGED, mapOf(
                OneTimePasswordChangedArgs.OTP to newOneTimePasscode
            ))
        }
    }

    override fun authenticationTokenStatus(data: Bundle?, error: PingID.PIDErrorDomain?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.AUTHENTICATION_TOKEN_STATUS, mapOf(
                AuthenticationTokenStatusArgs.DATA to data?.toMap(),
                AuthenticationTokenStatusArgs.ERROR to error.toString()
            ))
        }
    }

    //General Events
    override fun onServicePayloadReceivedWithStatusDone() {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.SERVICE_PAYLOAD_RECEIVED_WITH_STATUS_DONE, null)
        }
    }

    override fun onLogsSentToServer(status: PingID.PIDActionStatus?, supportId: String?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.LOGS_SENT_TO_SERVER, mapOf(
                LogsSentToServerArgs.ACTION_STATUS to status.toString(),
                LogsSentToServerArgs.SUPPORT_ID to supportId
            ))
        }
    }

    override fun onGeneralMessage(msg: String?) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.GENERAL_MESSAGE, mapOf(
                GeneralMessageArgs.MESSAGE to msg
            ))
        }
    }

    override fun onGooglePlayServicesStatusReceived(status: Int) {
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.GOOGLE_PLAY_SERVICES_STATUS_RECEIVED, mapOf(
                GooglePlayServicesStatusReceivedArgs.STATUS to status
            ))
        }
    }

    override fun onError(throwable: Throwable?, description: String?) {
        Log.d("PingEvent", "onError: Throwable: $throwable | $description")
        mainHandler.post {
            channel.invokeMethod(FlutterMethods.ERROR, mapOf(
                ErrorArgs.THROWABLE to throwable,
                ErrorArgs.DESCRIPTION to description
            ))
        }
    }
}